### PR TITLE
Use product keys as the name attribute

### LINKED ISSUE #1: Use product keys as the name attribute
Chef end-users don't care about our history of naming things poorly. We should rename the `chef_ingredient` resource's `package_name` attribute to `product` and perform mappings for them.

Here is a canonical list of products and associated keys/packages.

| Product | Product Key  | Package Name |
| ---------- | ------------------ | -------------------- |
| Chef Client | chef | chef |
| Chef Development Kit | chefdk| chefdk |
| Enterprise Chef (legacy) | private-chef | private-chef |
| Chef Server | chef-server | chef-server (versions < 12.0.0) <br/> chef-server-core (versions >= 12.0.0) |
| Chef Server High Availability addon | chef-ha | chef-ha |
| Chef Server Replication addon | chef-sync | chef-sync |
| Chef Server Reporting addon | reporting | opscode-reporting |
| Management Console | manage | opscode-manage (versions < 2.0.0) <br/> chef-manage (versions >= 2.0.0) |
| Push Jobs Server | push-jobs-server | opscode-push-jobs-server |
| Push Jobs Client | push-jobs-client | opscode-push-jobs-client (versions < 2.0.0) <br/> push-jobs-client (versions >= 2.0.0)  |
| Analytics Platform | analytics | opscode-analytics |
| Delivery | delivery | delivery |
| Delivery CLI | delivery-cli | delivery-cli |
| Supermarket | supermarket | supermarket |

Additionally the `chef_ingredient` resource should validate the `product` attribute and raise a friendly error if an invalid product key is provided.

### PR DESCRIPTION
**Pull request by jtimberman**
Fixes #1.

Implements a product matrix and product_lookup method that returns a
hash of product information based on the table in #1, with naming
conventions updated to upcoming product updates. It attempts to be
smart so users of the chef_ingredient resource don't need to know the
underlying details about our package naming, especially wrt old
package names.

The product_lookup method also validates that the chef-server package
is selected properly based on the available, supported versions (12
and 11), and fails if the version is specified as > 1.0 and < 11.0,
since we did not publish omnibus packages for those versions.

Add the version attribute to the omnibus_service resource so it can
properly look up the ctl command.

### CODE REVIEW

**Comment by jtimberman** [comment]
@chef-cookbooks/ociv

File: README.md [hunk 1]
```diff
@@ -37,7 +37,8 @@ A "chef ingredient" is the core package itself, or products or add-on components
 - `reconfigure` - Performs the `ctl reconfigure` command for the package.
 
 #### Properties
-- `package_name`: (name attribute) The name of the package. Should correspond to the published package names (chef-server-core, opscode-manage, etc).
+- `product_name`: (name attribute) The product name. See the `#product_matrix` method in `libraries/helpers.rb` for a list of valid product names. For example, `chef-server`, `analytics`, `delivery`, `manage`, etc.
+- `package_name`: The name of the package in the repository. Should correspond to the published package names (`chef-server-core`, `opscode-manage`, etc).
```
Comment (schisamo) [inline README.md]: What is the purpose of providing this attribute? It seems it should not be needed as the cookbook now ships with all logic needed to map a product name to a package name.

File: libraries/omnibus_service_resource.rb [hunk 2]
```diff
@@ -24,6 +24,7 @@ class OmnibusService < Chef::Resource::LWRPBase
 
       attribute :ctl_command, kind_of: String
       attribute :service_name, kind_of: String, regex: %r{[\w-]+/[\w-]+}, name_attribute: true
+      attribute :version, kind_of: String, default: '0.0.0'
```
Comment (schisamo) [inline libraries/omnibus_service_resource.rb]: Why is this needed?

**Comment by schisamo** [comment]
We should probably list the table from #1 in the README (__Product__ and __Product Key__ columns).

File: libraries/omnibus_service_resource.rb (hunk 2 above)
Comment (jtimberman) [inline libraries/omnibus_service_resource.rb]: `omnibus_service` uses the `chef_ctl_command` method to get the ctl command for the product, and it takes version as an argument. I can do something like this:

```ruby
    def chef_ctl_command(product)
      version = new_resource.respond_to?(:version) ? new_resource.version : '0.0.0'
      product_lookup(product, version)['ctl-command']
    end
```

**Comment by jtimberman** [comment]
I'll add that table as a separate file that isn't displayed on Supermarket, because it doesn't "do" GHFM tables well (or HTML tables).

File: README.md (hunk 1 above)
Comment (jtimberman) [inline README.md]: My thought was flexibility for transitioning, or in the event that a new package is created that isn't in the matrix (yet), without us having to cut a new chef-ingredient cookbook to update that.
